### PR TITLE
System header discovery should locate C++ standard library headers when using clang or gcc

### PR DESCRIPTION
**Pull request by wolfs**
See https://github.com/gradle/gradle-native/issues/281.

### CODE REVIEW

File: subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/gcc/GccPlatformToolProvider.java [hunk 1]
```diff
@@ -182,8 +198,17 @@ private String getPCHFileExtension() {
...
+        CommandLineToolSearchResult searchResult = toolSearchPath.locate(compiler.getToolType(), compiler.getExecutable());
+        String language = LANGUAGE_FOR_COMPILER.get(compilerType);
+        List<String> languageArgs = language == null ? Collections.<String>emptyList() : ImmutableList.of("-x", language);
+        return metadataProvider.getCompilerMetaData(searchResult.getTool(), languageArgs);
+    }
+
     @Override
-    public List<File> getSystemIncludes() {
+    public List<File> getSystemIncludes(ToolType compilerType) {
+        GccMetadata gccMetadata = getGccMetadata(compilerType);
```
Comment (ghale) [inline subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/gcc/GccPlatformToolProvider.java]: I'd love to see a unit test around this that proves that when I see the ToolType for each of the LANGUAGE_FOR_COMPILER types, I see a call to get a compiler metadata from `metadataProvider` with the correct languageArgs.  Seems like you should be able to stub everything out and do this pretty easily.

Comment (wolfs) [reply subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/gcc/GccPlatformToolProvider.java]: I added the test.

**Comment by wolfs** [comment]
@ghale Could you have another look?

**Comment by wolfs** [comment]
Hmm. There are some performance failures. I guess this is since we are now snapshotting even more include roots for C++ projects :(.

**Comment by ghale** [comment]
Changes look good.  Before we merge, we should investigate the performance test failure and decide whether to fix or rebaseline.

**Comment by wolfs** [comment]
@ghale I investigated and it looks like the performance regressions are due to more snapshotting. Which makes sense, since we have an new include root available and we find more actual includes (since we search those transitively). I would accept the regression.

**Pull request merged**